src/ucs/sys/socket.c

```diff
@@ -0,0 +1,193 @@
+{
+    ucs_status_t status;
+    int fd, ret;
+
+    ucs_strncpy_zero(if_req->ifr_name, if_name, sizeof(if_req->ifr_name));
+
+    status = ucs_socket_create(AF_INET, SOCK_STREAM, &fd);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    ret = ioctl(fd, request, if_req);
+    if (ret < 0) {
+        ucs_debug("ioctl(req=%lu, ifr_name=%s) failed: %m", request, if_name);
+        status = UCS_ERR_IO_ERROR;
+        goto out_close_fd;
+    }
+
+    status = UCS_OK;
+
+out_close_fd:
+    close(fd);
+out:
+    return status;
+}
+
+int ucs_netif_is_active(const char *if_name)
+{
+    ucs_status_t status;
+    struct ifreq ifr;
+
+    status = ucs_netif_ioctl(if_name, SIOCGIFADDR, &ifr);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    status = ucs_netif_ioctl(if_name, SIOCGIFFLAGS, &ifr);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    return (ifr.ifr_flags & IFF_UP) && (ifr.ifr_flags & IFF_RUNNING) &&
+           !(ifr.ifr_flags & IFF_LOOPBACK);
+}
+
+ucs_status_t ucs_socket_create(int domain, int type, int *fd_p)
+{
+    int fd = socket(domain, type, 0);
+    if (fd < 0) {
+        ucs_error("socket create failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    *fd_p = fd;
+    return UCS_OK;
+}
+
+ucs_status_t ucs_socket_connect(int fd, const struct sockaddr *dest_addr)
+{
+    char str[UCS_SOCKADDR_STRING_LEN];
+    int ret;
+
+    ret = connect(fd, dest_addr, ucs_sockaddr_sizeof(dest_addr));
+    if (ret < 0) {
+        if (errno == EINPROGRESS) {
+            ucs_debug("connect(fd=%d, dest_addr=%s): %m", fd,
+                      ucs_sockaddr_str(dest_addr, str, UCS_SOCKADDR_STRING_LEN));
+            return UCS_INPROGRESS;
+        }
+
+        ucs_error("connect(fd=%d, dest_addr=%s) failed: %m", fd,
+                  ucs_sockaddr_str(dest_addr, str, UCS_SOCKADDR_STRING_LEN));
+        return UCS_ERR_UNREACHABLE;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_socket_connect_nb_get_status(int fd)
+{
+    socklen_t conn_status_sz;
+    int ret, conn_status;
+
+    conn_status_sz = sizeof(conn_status);
+
+    ret = getsockopt(fd, SOL_SOCKET, SO_ERROR,
+                     &conn_status, &conn_status_sz);
+    if (ret < 0) {
+        ucs_error("getsockopt(fd=%d) failed to get SOL_SOCKET(SO_ERROR): %m", fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if ((conn_status == EINPROGRESS) || (conn_status == EWOULDBLOCK)) {
+        return UCS_INPROGRESS;
+    }
+
+    if (conn_status != 0) {
+        ucs_error("SOL_SOCKET(SO_ERROR) status on fd %d: %s", fd, strerror(conn_status));
+        return UCS_ERR_UNREACHABLE;
+    }
+
+    return UCS_OK;
+}
+
+size_t ucs_sockaddr_sizeof(const struct sockaddr *addr)
+{
+    switch (addr->sa_family) {
+    case AF_INET:
+        return sizeof(struct sockaddr_in);
+    case AF_INET6:
+        return sizeof(struct sockaddr_in6);
+    default:
+        ucs_error("unknown address family: %d", addr->sa_family);
+        return 0ULL;
```

Comment:
AFAIK there is no real need for ULL, it would be casted automatically from int to size_t
@brminich